react-redux/index.d.ts

```diff
@@ -6,102 +6,119 @@
 
-type ComponentClass<P> = React.ComponentClass<P>;
-type StatelessComponent<P> = React.StatelessComponent<P>;
-type ReactNode = React.ReactNode;
-type Store<S> = Redux.Store<S>;
-type Dispatch<S> = Redux.Dispatch<S>;
-type ActionCreator<A> = Redux.ActionCreator<A>;
-
-interface ComponentDecorator<TOriginalProps, TOwnProps> {
-    (component: ComponentClass<TOriginalProps> | StatelessComponent<TOriginalProps>): ComponentClass<TOwnProps>;
-}
+export = ReactRedux;
 
-/**
- * Decorator that infers the type from the original component
- *
- * Can't use the above decorator because it would default the type to {}
- */
-export interface InferableComponentDecorator {
-    <P, TComponentConstruct extends (ComponentClass<P> | StatelessComponent<P>)>(component: TComponentConstruct): TComponentConstruct;
-}
+declare namespace ReactRedux {
 
-/**
- * Connects a React component to a Redux store.
- *
- * - Without arguments, just wraps the component, without changing the behavior / props
- *
- * - If 2 params are passed (3rd param, mergeProps, is skipped), default behavior
- * is to override ownProps (as stated in the docs), so what remains is everything that's
- * not a state or dispatch prop
- *
- * - When 3rd param is passed, we don't know if ownProps propagate and whether they
- * should be valid component props, because it depends on mergeProps implementation.
- * As such, it is the user's responsibility to extend ownProps interface from state or
- * dispatch props or both when applicable
- *
- * @param mapStateToProps
- * @param mapDispatchToProps
- * @param mergeProps
- * @param options
- */
-export declare function connect(): InferableComponentDecorator;
-
-export declare function connect<TStateProps, TDispatchProps, TOwnProps>(
-    mapStateToProps: FuncOrSelf<MapStateToProps<TStateProps, TOwnProps>>,
-    mapDispatchToProps?: FuncOrSelf<MapDispatchToPropsFunction<TDispatchProps, TOwnProps> | MapDispatchToPropsObject>
-): ComponentDecorator<TStateProps & TDispatchProps, TOwnProps>;
-
-export declare function connect<TStateProps, TDispatchProps, TOwnProps>(
-    mapStateToProps: FuncOrSelf<MapStateToProps<TStateProps, TOwnProps>>,
-    mapDispatchToProps: FuncOrSelf<MapDispatchToPropsFunction<TDispatchProps, TOwnProps> | MapDispatchToPropsObject>,
-    mergeProps: MergeProps<TStateProps, TDispatchProps, TOwnProps>,
-    options?: Options
-): ComponentDecorator<TStateProps & TDispatchProps, TOwnProps>;
-
-type FuncOrSelf<T> = T | (() => T);
-
-interface MapStateToProps<TStateProps, TOwnProps> {
-    (state: any, ownProps?: TOwnProps): TStateProps;
-}
+    type ComponentType<P> = React.ComponentClass<P> | React.StatelessComponent<P>;
 
-interface MapDispatchToPropsFunction<TDispatchProps, TOwnProps> {
-    (dispatch: Dispatch<any>, ownProps?: TOwnProps): TDispatchProps;
-}
+    interface ComponentDecorator<M, P> {
+        <TargetClass extends ComponentType<M>>(component: TargetClass): TargetClass & ConnectClass<P, TargetClass>;
```

Comment:
Hm, interesting. If your first comment is consistent, then that would be awesome! In the typescript playground, I put together [this example](https://www.typescriptlang.org/play/#src=interface%20I1%20%7B%0D%0A%20%20foo%3A%20string%3B%0D%0A%20%20bar%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20I2%20%7B%0D%0A%20%20foo%3A%20number%3B%0D%0A%20%20baz%3A%20number%3B%0D%0A%7D%0D%0A%0D%0Aclass%20Class%3CP%3E%20%7B%0D%0A%20%20props%3A%20P%3B%0D%0A%7D%0D%0A%0D%0Atype%20MixedClass%20%3D%20Class%3CI1%3E%20%26%20Class%3CI2%3E%3B%0D%0A%0D%0Alet%20m%3A%20MixedClass%3B%0D%0A%0D%0A):

```ts
interface I1 {
  foo: string;
  bar: string;
}

interface I2 {
  foo: number;
  baz: number;
}

class Class<P> {
  props: P;
}

type MixedClass = Class<I1> & Class<I2>;

let m: MixedClass;
```

If you open it up and poke around `m` you can see what it autocompletes: it seems to be doing a recursive intersection on the properties, which makes sense to me. I'm trying to figure out what it is about this case that's causing it to apparently do the right thing, weird though the right thing is.